File: riveredge-backend/src/apps/kuaizhizao/schemas/production_daily.py
"""生产日报 schemas。"""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class ProductionDailyFieldDef(BaseModel):
    key: str = Field(..., min_length=1, max_length=64)
    label: str = Field(..., min_length=1, max_length=100)
    type: str = Field(default="text", description="text/number/textarea/select")
    required: bool = False
    options: Optional[List[str]] = None


class ProductionDailyTemplateCreate(BaseModel):
    template_code: Optional[str] = Field(None, max_length=50)
    template_name: str = Field(..., max_length=200)
    description: Optional[str] = None
    field_schema: List[ProductionDailyFieldDef] = Field(default_factory=list)
    sort_order: int = 0
    is_active: bool = True


class ProductionDailyTemplateUpdate(BaseModel):
    template_name: Optional[str] = Field(None, max_length=200)
    description: Optional[str] = None
    field_schema: Optional[List[ProductionDailyFieldDef]] = None
    sort_order: Optional[int] = None
    is_active: Optional[bool] = None


class ProductionDailyTemplateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: str
    template_code: str
    template_name: str
    description: Optional[str] = None
    field_schema: List[Any] = Field(default_factory=list)
    sort_order: int = 0
    is_active: bool = True
    is_system: bool = False
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None


class ProductionDailyTemplateListResponse(BaseModel):
    items: List[ProductionDailyTemplateResponse]
    total: int


class ProductionDailyReportCreate(BaseModel):
    template_id: int
    report_date: date
    team_name: Optional[str] = None
    shift_name: Optional[str] = None
    workshop_name: Optional[str] = None
    plant_name: Optional[str] = None
    field_values: dict[str, Any] = Field(default_factory=dict)
    remarks: Optional[str] = None
    submit: bool = False


class ProductionDailyReportUpdate(BaseModel):
    report_date: Optional[date] = None
    team_name: Optional[str] = None
    shift_name: Optional[str] = None
    workshop_name: Optional[str] = None
    plant_name: Optional[str] = None
    field_values: Optional[dict[str, Any]] = None
    remarks: Optional[str] = None


class ProductionDailyReportResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: str
    code: str
    template_id: int
    template_code: str
    template_name: str
    report_date: date
    team_name: Optional[str] = None
    shift_name: Optional[str] = None
    workshop_name: Optional[str] = None
    plant_name: Optional[str] = None
    field_values: dict[str, Any] = Field(default_factory=dict)
    status: str
    submitted_at: Optional[datetime] = None
    remarks: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None


class ProductionDailyReportListResponse(BaseModel):
    items: List[ProductionDailyReportResponse]
    total: int

File: riveredge-frontend/src/apps/kuaizhizao/services/production-daily.ts
import { apiRequest } from '../../../services/api';

const TPL = '/apps/kuaizhizao/production-daily-templates';
const RPT = '/apps/kuaizhizao/production-daily-reports';

export type ProductionDailyFieldDef = {
  key: string;
  label: string;
  type: 'text' | 'number' | 'textarea' | 'select';
  required?: boolean;
  options?: string[];
};

export type ProductionDailyTemplate = {
  id: number;
  uuid: string;
  template_code: string;
  template_name: string;
  description?: string | null;
  field_schema: ProductionDailyFieldDef[];
  sort_order: number;
  is_active: boolean;
  is_system: boolean;
  created_by_name?: string | null;
  updated_by_name?: string | null;
};

export type ProductionDailyReport = {
  id: number;
  uuid: string;
  code: string;
  template_id: number;
  template_code: string;
  template_name: string;
  report_date: string;
  team_name?: string | null;
  shift_name?: string | null;
  workshop_name?: string | null;
  plant_name?: string | null;
  field_values: Record<string, unknown>;
  status: string;
  submitted_at?: string | null;
  remarks?: string | null;
  created_by_name?: string | null;
  updated_by_name?: string | null;
};

function unwrapList<T>(res: unknown): { items: T[]; total: number } {
  if (Array.isArray(res)) return { items: res as T[], total: res.length };
  const r = res as Record<string, unknown>;
  const items = (r.items ?? r.data ?? []) as T[];
  return { items: Array.isArray(items) ? items : [], total: Number(r.total ?? items.length) };
}

export const productionDailyTemplateApi = {
  list: async (params?: Record<string, unknown>) =>
    unwrapList<ProductionDailyTemplate>(await apiRequest(TPL, { method: 'GET', params })),
  get: async (id: number) =>
    (await apiRequest(`${TPL}/${id}`, { method: 'GET' })) as ProductionDailyTemplate,
  create: async (data: Partial<ProductionDailyTemplate>) =>
    (await apiRequest(TPL, { method: 'POST', data })) as ProductionDailyTemplate,
  update: async (id: number, data: Partial<ProductionDailyTemplate>) =>
    (await apiRequest(`${TPL}/${id}`, { method: 'PUT', data })) as ProductionDailyTemplate,
  remove: async (id: number) => apiRequest(`${TPL}/${id}`, { method: 'DELETE' }),
  ensureDefaults: async () =>
    apiRequest(`${TPL}/ensure-defaults`, { method: 'POST' }),
};

export const productionDailyReportApi = {
  list: async (params?: Record<string, unknown>) =>
    unwrapList<ProductionDailyReport>(await apiRequest(RPT, { method: 'GET', params })),
  get: async (id: number) =>
    (await apiRequest(`${RPT}/${id}`, { method: 'GET' })) as ProductionDailyReport,
  create: async (data: Record<string, unknown>) =>
    (await apiRequest(RPT, { method: 'POST', data })) as ProductionDailyReport,
  update: async (id: number, data: Record<string, unknown>) =>
    (await apiRequest(`${RPT}/${id}`, { method: 'PUT', data })) as ProductionDailyReport,
  submit: async (id: number) =>
    (await apiRequest(`${RPT}/${id}/submit`, { method: 'POST' })) as ProductionDailyReport,
  remove: async (id: number) => apiRequest(`${RPT}/${id}`, { method: 'DELETE' }),
};
